wasm/wazeroir/interpreter.go

```diff
@@ -0,0 +1,1502 @@
+			// so just eliminate the ops.
+			continue
+		case *OperationExtend:
+			if o.Signed {
+				op.b1 = 1
+			}
+		default:
+			return nil, fmt.Errorf("unreachable: a bug in wazeroir interpreter")
+		}
+		ret.body = append(ret.body, op)
+	}
+
+	if len(onLabelAddressResolved) > 0 {
+		keys := make([]string, 0, len(onLabelAddressResolved))
+		for key := range onLabelAddressResolved {
+			keys = append(keys, key)
+		}
+		return nil, fmt.Errorf("labels are not defined: %s", strings.Join(keys, ","))
+	}
+	return ret, nil
+}
+
+// Implements wasm.Engine for interpreter.
+func (it *interpreter) Call(f *wasm.FunctionInstance, args ...uint64) (returns []uint64, err error) {
+	prevFrameLen := len(it.frames)
+	defer func() {
+		if v := recover(); v != nil {
+			// TODO: include stack trace in the error message.
+			if buildoptions.IsDebugMode {
+				debug.PrintStack()
+			}
+			it.frames = it.frames[:prevFrameLen]
+			err2, ok := v.(error)
+			if ok {
+				err = fmt.Errorf("runtime error: %w", err2)
+			} else {
+				err = fmt.Errorf("runtime error: %v", v)
+			}
+		}
+	}()
+
+	g, ok := it.functions[f]
+	if !ok {
+		err = fmt.Errorf("function not compiled")
+		return
+	}
+
+	if g.hostFn != nil {
+		it.callHostFunc(g, args...)
+	} else {
+		for _, arg := range args {
+			it.push(arg)
+		}
+		it.callNativeFunc(g)
+	}
+	returns = make([]uint64, len(f.Signature.ReturnTypes))
+	for i := range returns {
+		returns[len(returns)-1-i] = it.pop()
+	}
+	return
+}
+
+func (it *interpreter) callHostFunc(f *interpreterFunction, args ...uint64) {
+	tp := f.hostFn.Type()
+	in := make([]reflect.Value, tp.NumIn())
+	for i := len(in) - 1; i >= 1; i-- {
+		val := reflect.New(tp.In(i)).Elem()
+		raw := it.pop()
+		kind := tp.In(i).Kind()
+		switch kind {
+		case reflect.Float64, reflect.Float32:
+			val.SetFloat(math.Float64frombits(raw))
+		case reflect.Uint32, reflect.Uint64:
+			val.SetUint(raw)
+		case reflect.Int32, reflect.Int64:
+			val.SetInt(int64(raw))
+		}
+		in[i] = val
+	}
+
+	val := reflect.New(tp.In(0)).Elem()
+	var memory *wasm.MemoryInstance
+	if len(it.frames) > 0 {
+		memory = it.frames[len(it.frames)-1].f.moduleInstance.Memory
+	}
+	val.Set(reflect.ValueOf(&wasm.HostFunctionCallContext{Memory: memory}))
+	in[0] = val
+
+	frame := &interpreterFrame{f: f}
+	it.pushFrame(frame)
+	for _, ret := range f.hostFn.Call(in) {
+		switch ret.Kind() {
+		case reflect.Float64, reflect.Float32:
+			it.push(math.Float64bits(ret.Float()))
+		case reflect.Uint32, reflect.Uint64:
+			it.push(ret.Uint())
+		case reflect.Int32, reflect.Int64:
+			it.push(uint64(ret.Int()))
+		default:
+			panic("invalid return type")
+		}
+	}
+	it.popFrame()
+}
+
+func (it *interpreter) callNativeFunc(f *interpreterFunction) {
+	frame := &interpreterFrame{f: f}
+	memoryInst := frame.f.moduleInstance.Memory
+	globals := frame.f.moduleInstance.Globals
+	tables := f.moduleInstance.Tables
+	it.pushFrame(frame)
+	bodyLen := uint64(len(frame.f.body))
+	for frame.pc < bodyLen {
```

Comment:
add a TODO or mental note to comment either on the constants or each case block what the operation implies in terms of how many if any args are popped and what if anything we expect to push, if there are any expected handled errors (ex overflow returns zero value or whatever). Main idea isn't to copy the spec, just a one max 2 line overview per non-trivial op